M2-5096: convert remaining popups to banners

farmerpaul opened the pull request:
### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-5096](https://mindlogger.atlassian.net/browse/M2-5096) (parent ticket [M2-4823](https://mindlogger.atlassian.net/browse/M2-4823))

This PR converts the remaining popups (as selected from the [popups audit](https://mindlogger.atlassian.net/browse/M2-4968)) into banners. In doing so, added support for passing props to banners (as not all types of banners have static text). Also, some minor copy changes were made to error popups (which continue to be popups, as per audit).

The PR also introduces a utility for testing the presence of a banner (`expectBanner`) to make it easier to test for the presence of a banner (see [note below](#expect-banner)).

It will likely be easier to review this PR [commit-by-commit](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1486/commits). (FYI: Commit messages reference table rows by item # in the [popups audit](https://mindlogger.atlassian.net/browse/M2-4968).)

### 📸 Screenshots

https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/eaa8a151-4411-4f1a-9c73-f2bbebd54d63

### 🪤 Peer Testing

Follow testing steps below to confirm behaviour:

| Testing steps | Expected banner text | Duration |
| - | - | - |
| Edit applet > Applet settings > Report configuration > Save > Save anyway | `Changes saved` | 3.5s |
| Edit applet > Edit About Applet Page > Upload Image > Select image 30+ MB image | `File is too big. Please upload a file less than 25 MB.` | 5s |
| Edit applet > Edit About Applet Page > Upload Image > Select BMP image | `Image format must be JPEG or PNG.` | 5s |
| Edit applet > Edit About Applet Page > Upload Audio > Select OGG audio | `Incorrect audio format` | 5s |
| Edit applet > Edit About Applet Page > Upload Audio > Select video 200+ MB video | `File is too big. Please upload a file less than 150 MB.` | 5s |
| Edit applet > Applet settings > Data retention > Save | `Changes saved` | 3.5s |
| Dashboard > Delete an applet > Enter password > Delete | `Applet was deleted` | 3.5s |
| Dashboard > Transfer ownership > Enter another user email > Confirm | `Your request was successfully sent to <email>. Please wait for receiver to accept your request.` | 7s |
| Dashboard > Choose "View Users" for an applet with multiple managers > Managers > Edit another manager's access > Save | `Changes saved` | 3.5s |
| Dashboard > Choose "View Users" for an applet with respondents > Respondents > Edit a respondent > Save | `Changes saved` | 3.5s |

### ✏️ Notes

<a id="testing"></a>
#### Missing ability to test specific popup
⚠️ Although I've converted the below popups to banners as well in this PR (as per requirements), I have not yet figured out **how to test these two**. The code lives in `PublishConcealAppletPopup.tsx`. If someone at ScienceSoft would be able to provide insight into this, that would be great!

| Testing steps | Expected banner text | Duration |
| - | - | - |
| ❓ | `The applet <applet name> has been concealed successfully` | Persistent |
| ❓ | `The applet <applet name> has been published successfully`* | Persistent |

* Note that this second popup is **not** the same as the `Applet <applet name> has been successfully saved and published.` popup, which has _not_ been converted to a banner as part of this PR. That one is expected to be done as part of [M2-5089](https://mindlogger.atlassian.net/browse/M2-5089).

<a id="expect-banner"></a>
#### `expectBanner` utility
In the tests, my hope was to test for the presence of banners by matching against the rendered text found within them, rather than using Redux state directly. However, banners aren't rendered in our tests unless the `<Banners />` component is included somewhere. Hypothetically, `<Banners />` could be embedded within our `renderWithProviders` and `renderHookWithProviders` utilities, however I tried that, and it caused circular dependency issues that surfaced either during Jest execution, or the app, or even both. Rather than spend an undue amount of time attempting to resolve those, I opted to stick with comparing against Redux state directly to check for presence of banners.

[M2-4823]: https://mindlogger.atlassian.net/browse/M2-4823?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[M2-5089]: https://mindlogger.atlassian.net/browse/M2-5089?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

farmerpaul committed:
M2-5096: minor Banner refactor for custom props

Allows the component displaying the banner to pass custom props to
different banner types that require arguments (e.g. custom onClose
handler, dynamic banner messages).

Also reduce dependencies between Redux state and component library to
discourage circular dependency issues (particularly w/ Jest) by
creating an enum of the different banner keys, and associating those
keys with the actual components within the Banners component.

Also add support for testID to Banner component.

farmerpaul committed:
M2-5096: Update copy: save report error (item #2)

Switching to generic `saveError` key for copy.

farmerpaul committed:
M2-5096: SuccessPopup ⇒ SaveSuccessBanner (item #4)

Convert ReportConfigSetting's SucessPopup into shared SaveSuccessBanner
to be used in multiple places in the app. Fix ReportConfigSetting tests.

Also update VersionWarningBanner's tests to be more comprehensive,
aligning with SaveSuccessBanner.

farmerpaul committed:
M2-5096: size/format error popups ⇒ banners (item #5)

Convert both FileSizeExceededBanner and IncorrectFileBanner to banners.

farmerpaul committed:
M2-5096: data retention success ⇒ banner (item #7)

farmerpaul committed:
M2-5096: applet delete success ⇒ banner (item #8)

Reuse existing success banner & add ability to override default message.

farmerpaul committed:
M2-5096: publish/conceal success ⇒ banner (item #10)

This banner persists (duration === null) as per requirements.

farmerpaul committed:
M2-5096: transfer ownership ✔️ ⇒ banner (item #12)

Convert transfer ownership success popup to a banner. Also fix test
regression introduced by Delete Applet banner a couple commits ago.

farmerpaul committed:
M2-5096: edit access success ⇒ banner (item #13)

farmerpaul committed:
M2-5096: edit respondent success ⇒ banner (item #14)

farmerpaul committed:
M2-5096: create banner utility for tests

Reduce repeated code and make testing for presence of banners easier.

farmerpaul committed:
M2-5096: use generic copy for error (item #6)

sabrichu commented on src/shared/utils/expectBanner.ts:
Nice, thanks for adding!

sabrichu commented on src/shared/state/Banners/Banners.schema.ts:
```diff
@@ -1,13 +1,16 @@
+import { BannerProps } from 'shared/components/Banners/Banner/Banner.types';
 
-export const BannerComponents = {
+export enum BannerType {
```
Not a blocking comment, but is the expectation that with any new banner added to the site, we would add a new type here? I would expect these banner types to be much more generic (`SuccessBanner`, `WarningBanner`, `ErrorBanner`, etc) with the children supplied via `bannerProps`, and not being much more opinionated than that...

...but then again, those wrapper components also wouldn't be doing much more than pre-configuring some values for `severity` `duration`. Anyway, just food for thought. My PRs were going to reuse the `SaveSuccessBanner` for save success and duplicate success

farmerpaul replied:
I've also been reusing `SaveSuccessBanner` for most success situations. You're right that it is basically just props that are changing (with `children` being one of those props), but I suppose the benefit of having distinct components is cases where the content is more complex and that banner is called from multiple locations, just to offer a level of reusability. (Which, as I say that, it now occurs to me I probably should have done with the "transfer ownership" banner! Maybe I'll make a change there…)

sabrichu replied:
Got it, that seems reasonable!

sabrichu reviewed (commented):
Still looking through it all but wowza, great work! This was a lot to comb through/it's not easy maneuvering a mix of legacy approaches and components

Have an initial comment...

sabrichu commented on src/shared/components/Uploader/Uploader.tsx:
```diff
@@ -197,14 +201,51 @@ export const Uploader = ({
-  const hasSizeError = error === UploadFileError.Size;
-  const hasFormatError = error === UploadFileError.Format;
-  const hasDimensionsError = error === UploadFileError.Dimensions;
+  const [hasSizeError, hasFormatError, hasDimensionsError] = useMemo(
```
Curious why `useMemo` is needed here

farmerpaul replied:
Good question. Neither of my `useMemo`s here are particularly warranted. I think I was trying to troubleshoot reactive behaviour during testing and was misguided into thinking this would solve it. I'll get rid of these.

sabrichu replied:
I have scattered many a `setTimeout` for the exact same reason

alex-i-e commented on src/modules/Builder/features/ReportConfigSetting/ReportConfigSetting.tsx:
```diff
@@ -382,10 +382,19 @@ export const ReportConfigSetting = ({
   };
 
   useEffect(() => {
```
I guess we don't need local state `successBannerVisible/setSuccessBannerVisible` anymore, since you could retrieve this info from `banners` store (via check if `SaveSuccessBanner` is in the list). Moreover, if you move `dispatch(banners.actions.addBanner({key: 'SaveSuccessBanner', ...` into a callback you could get rid of such places where used the code
```
 () => {
      setSuccessBannerVisible(true);
    },
```
and at the same time no need `setSuccessBannerVisible(false);`

alex-i-e commented on src/modules/Dashboard/features/Applet/Popups/DeletePopup/DeletePopup.tsx:
```diff
@@ -56,6 +56,21 @@ export const DeletePopup = ({ onCloseCallback, 'data-testid': dataTestid }: Dele
   };
 
+  useEffect(() => {
+    if (activeModal === Modals.Confirmation) {
```
let's get rid of `useEffect` that could be replaced with a callback. it's a way more intuitive and much improve readability of the code, simplify writing the tests and debugging.

alex-i-e reviewed (changes requested):
...left the comments in the code

alex-i-e commented on src/modules/Dashboard/features/Applet/Popups/TransferOwnershipPopup/TransferOwnershipPopup.tsx:
```diff
@@ -37,16 +33,38 @@ export const TransferOwnershipPopup = () => {
 
   useEffect(() => {
-    if (!emailTransfered) return;
+    if (!emailTransferred) return;
```
if it'll not take a long time let's get rid of `useEffect` that could be replaced with a callback as I guess into the places where triggered `setEmailTransferred(true)`

alex-i-e commented on src/shared/components/Banners/Banners.tsx:
```diff
@@ -1,21 +1,29 @@
-            <BannerComponent onClose={() => dispatch(banners.actions.removeBanner({ key }))} />
+            <BannerComponent
+              {...bannerProps}
+              onClose={() => {
```
As a common rule, we use an anonymous function into the html as a callback handler for cases when it includes one callback/action. In this case I would move the logic into a separate function. It reduces html code and improves readability. But feel free discuss it of course.

alex-i-e commented on src/shared/components/Uploader/Uploader.tsx:
```diff
@@ -197,14 +201,51 @@ export const Uploader = ({
 
   const fileName = imageField?.split('/').at(-1) || image?.name || '';
 
+  useEffect(() => {
```
Here, as well let's get rid of `useEffect` that could be replaced with a callback into the places where `error` set to `Size` and `Format` accordingly (in row 110 and 111) , I suppose, when the condition `isPrimaryUiType === false`

alex-i-e commented on src/shared/features/AppletSettings/TransferOwnershipSetting/TransferOwnershipSetting.tsx:
```diff
@@ -1,39 +1,60 @@
 
   useEffect(() => {
-    if (!emailTransfered) return;
+    if (!emailTransferred) return;
```
if it'll not take a long time let's get rid of `useEffect` that could be replaced with a callback as I guess into the places where triggered `setEmailTransferred(true)`

alex-i-e commented:
Btw 👍 , nice to see good test coverage, thank you!

farmerpaul committed:
M2-5096: remove effects & memos in Uploader

farmerpaul committed:
M2-5096: transfer ownership banner improvements

Created a reusable banner component for transferring ownership since
it's used in multiple places. Also simplify logic by removing effects.

farmerpaul committed:
M2-5096: remove ReportConfigSetting effect

IhorNesterenko99 commented on src/modules/Dashboard/features/Applet/Popups/DeletePopup/DeletePopup.tsx:
```diff
@@ -90,18 +105,8 @@ export const DeletePopup = ({ onCloseCallback, 'data-testid': dataTestid }: Dele
         </Modal>
       );
     case Modals.Confirmation:
```
Could you please remove this case from switch statement and from Modals enum?

farmerpaul committed:
M2-5096: remove DeletePopup effect

farmerpaul committed:
M2-5096: convert anon function to callback

farmerpaul commented:
Appreciate your reviews @alex-i-e and @IhorNesterenko99! I agree with all your suggestions and have implemented them, so please take another look as soon as you have a moment.

I'm also curious if either of you know the answer to my [question above](https://github.com/ChildMindInstitute/mindlogger-admin/pull/1486#testing) about testability of the "applet concealed/published" functionality?

alex-i-e commented on src/modules/Builder/features/ReportConfigSetting/ReportConfigSetting.tsx:
```diff
@@ -104,7 +103,7 @@ export const ReportConfigSetting = ({
   const { execute: postReportConfig } = useAsync(
     postReportConfigApi,
     () => {
```
💅: here and below no need an anonymous function wrapper, right?
```
() => {
   handleSuccess();
}
```
use just the reference `handleSuccess` on the function

farmerpaul replied:
Agreed, and it's what I was going to do, but it required a little rearranging of code. Event handler now has to be defined early, separately from where the other event handlers live. I'm fine with that if you are, so pushing this change!

alex-i-e replied:
I'm ok with it, thank you!

farmerpaul committed:
M2-5096: unwrap callback from anon function

IhorNesterenko99 commented on src/modules/Builder/components/ToggleItemContainer/ToggleItemContainer.styles.ts:
```diff
@@ -28,15 +28,18 @@ export const StylesTitleWrapper = styled(StyledFlexTopCenter, shouldForwardProp)
     open,
```
It appears that the commits from our previous work, which have already been merged into the development branch, are now showing in your pull request. Could you please double-check?

farmerpaul replied:
How curious, yes, I'm seeing that above:

<img width="916" alt="Screenshot 2024-02-15 at 10 21 00 AM" src="https://github.com/ChildMindInstitute/mindlogger-admin/assets/1282772/b2a33ed0-318a-4540-9651-727a24423abb">

I have no idea how this happened; all I've been doing is rebasing on `develop`. I'll fix it, thanks for the callout.

IhorNesterenko99 commented on src/modules/Dashboard/features/Respondents/Popups/EditRespondentPopup/EditRespondentPopup.tsx:
```diff
@@ -76,53 +88,46 @@ export const EditRespondentPopup = ({
-      buttonText={t(isSuccessVisible ? 'ok' : 'save')}
-      hasSecondBtn={!isSuccessVisible}
+      buttonText={t('save')}
+      hasSecondBtn={true}
```
```suggestion
      hasSecondBtn
```

IhorNesterenko99 commented on src/shared/components/Banners/Banners.tsx:
```diff
@@ -1,21 +1,32 @@
 import { Collapse } from '@mui/material';
 import { TransitionGroup } from 'react-transition-group';
+import { useCallback } from 'react';
```
Could you please relocate the import from React to the top, placing it before the other imports?

IhorNesterenko99 commented on src/shared/components/Banners/Banners.tsx:
```diff
@@ -1,21 +1,32 @@
   const dispatch = useAppDispatch();
   const data = banners.useData();
 
+  const handleClose = useCallback(
```
Would it be beneficial to use useCallback here? What advantages would this decision provide?

alex-i-e replied:
Per my understanding, to not rerender `<BannerComponent ... />` which uses it this property. 

But, in this case, to prevent redundant renders of all `<BannerComponent...` from the list, it should have a bit another location and sequence of contexts for the property `handleClose`:

1). why we need to think about `useCallback/useMemo/memo` when we don't have a huge calculations/bad performance in descendants or in current component? I suppose it looks like overengineering here and in other similar places.

2.1). it should be outside of the component `Banners` to avoid the generating of new reference of the function each time when `const data = banners.useData();` is updated
```
const handlePureClose = (dispatch: AppDispatch, { key, bannerProps }: BannerPayload) => () => {
  dispatch(banners.actions.removeBanner({ key }));
  bannerProps?.onClose?.();
}
```
2.2). not to generate new anonymous function when banners data is updated:
```
<BannerComponent {...bannerProps} onClose={handlePureClose({ key, bannerProps })} />
```

IhorNesterenko99 commented on src/shared/components/Banners/SaveSuccessBanner/SaveSuccessBanner.tsx:
```diff
@@ -0,0 +1,13 @@
+  const { t } = useTranslation('app');
+
+  return (
+    <Banner duration={3500} severity="success" data-testid="save-success-banner" {...props}>
```
Could you please create a variable for the duration and place it in the SaveSuccessBanner.const.ts file?

IhorNesterenko99 commented on src/shared/components/Banners/TransferOwnershipSuccessBanner/TransferOwnershipSuccessBanner.tsx:
```diff
@@ -0,0 +1,22 @@
+
+export const TransferOwnershipSuccessBanner = ({ children, email, ...props }: BannerProps) => (
+  <Banner
+    duration={7000}
```
Could you please create a variable for the duration and place it in the TransferOwnershipSuccessBanner.const.ts file?

IhorNesterenko99 commented on src/shared/components/FormComponents/EditorController/EditorController.tsx:
```diff
@@ -19,53 +19,68 @@ export const EditorController = <T extends FieldValues>({
-  const [fileSizeExceeded, setFileSizeExceeded] = useState<number | null>(null);
-  const [incorrectFileFormat, setIncorrectFileFormat] = useState<MediaType | null>(null);
+
+  const onFileSizeExceeded = useCallback(
```
I'm curious, what are the benefits of using useCallback here and in the next function?

alex-i-e replied:
I guess to not re-render the huge component `<Editor` (use the old reference in properties when parent component is re-rendered). 
I think it makes sense, if it has a performance issues. Otherwise, I would consider it as a overengineering.

IhorNesterenko99 commented:
Hi @farmerpaul!
Thank you for taking our comments into consideration. 
Currently, the ability to Publish/Conceal requires the user to have a super-admin role. However, development of this functionality has been paused. Unfortunately, this means you won't be able to test it at this time.
Please take a look at the corresponding User Story:
https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/189104593/AUS-4.1.4.11+-+Applet+settings+tab+of+the+Applet+page+Publish+Conceal+Applet+for+MindLogger+System+Admin+Only

alex-i-e reviewed (approved):
LGTM

IhorNesterenko99 commented:
Additionally, I suggest reducing the code size for review by breaking down large tasks into smaller parts. This is something to consider for future pull requests.

farmerpaul committed:
M2-5096: PR review change requests

Convert useCallback into pure function. Also some small tidyups such as
moving constants into `.const.ts` files.
